### PR TITLE
node status api

### LINKED ISSUE #5762: NodeStatus API Endpoint for checking node health
Initially, it would be used to confirm that node is in a good state before producing a message.

Checks proposed:
- the sync status: how far out of sync the node is.
   - Use the difference between the current time and genesis to figure out which epoch we should be at.
- peering status: do we have more than 5 peers with positive pubsub scores?
   - i.e. are we confident that if we send this message now, it will be properly propagated to the network?

### PR DESCRIPTION
**Pull request by vyzo**
Closes #5762

### CODE REVIEW

**Comment by vyzo** [comment]
@magik6k what do you want to do with the chain health thing?

File: node/impl/full.go [hunk 1]
```diff
@@ -37,4 +40,40 @@ func (n *FullNodeAPI) CreateBackup(ctx context.Context, fpath string) error {
...
+	}
+
+	scoreMap := make(map[peer.ID]float64)
+	for _, score := range scores {
+		scoreMap[score.ID] = score.Score.Score
+	}
+
+	for _, pi := range pis {
+		score := scoreMap[pi.ID]
+		if score > -1000 {
```
Comment (raulk) [inline node/impl/full.go:71]: Where is this threshold coming from?

File: node/impl/full.go [hunk 2]
```diff
@@ -37,4 +40,40 @@ func (n *FullNodeAPI) CreateBackup(ctx context.Context, fpath string) error {
...
+	}
+
+	scores, err := n.NetPubsubScores(ctx)
+	if err != nil {
+		return status, err
+	}
+
+	scoreMap := make(map[peer.ID]float64)
+	for _, score := range scores {
+		scoreMap[score.ID] = score.Score.Score
```
Comment (raulk) [inline node/impl/full.go:66]: This map could be large, and it looks like too much detail for most use cases. Can we put it behind a flag?

File: node/impl/full.go [hunk 3]
```diff
@@ -37,4 +40,40 @@ func (n *FullNodeAPI) CreateBackup(ctx context.Context, fpath string) error {
...
+	}
+
+	for _, pi := range pis {
+		score := scoreMap[pi.ID]
+		if score > -1000 {
+			status.PeerStatus.PeersToPublish++
```
Comment (raulk) [inline node/impl/full.go:72]: Can we break down peers to publish by topic? The message management CLI cares about mpool, not about blocks (I think?).

Comment (vyzo) [reply node/impl/full.go:72]: I guess we could list peers by topic and only use those; but in practice all of them subscribe to the same topics as us.

File: node/impl/full.go (hunk 2 above)
Comment (vyzo) [reply node/impl/full.go:66]: Hrm, maybe; it's kind of important however, at least for the logical health checks.

File: node/impl/full.go (hunk 1 above)
Comment (vyzo) [reply node/impl/full.go:71]: our score configuration, no good way to reference it.

File: node/impl/full.go (hunk 3 above)
Comment (vyzo) [reply node/impl/full.go:72]: We can also break it into two, peers to publish blocks and peers to publish messages.

File: node/impl/full.go (hunk 1 above)
Comment (raulk) [reply node/impl/full.go:71]: Can we create a constant in right place and reference it from here? I'm afraid that if we change the threshold at origin, we will forget to update it here.

File: node/impl/full.go (hunk 2 above)
Comment (vyzo) [reply node/impl/full.go:66]: also, how large is large? It won't be more than 1k in practice.

Comment (vyzo) [reply node/impl/full.go:66]: we don't report specific peers, just the count of peers above the threshold.

File: node/impl/full.go (hunk 1 above)
Comment (vyzo) [reply node/impl/full.go:71]: probably.

File: node/impl/full.go [hunk 4]
```diff
@@ -37,4 +40,40 @@ func (n *FullNodeAPI) CreateBackup(ctx context.Context, fpath string) error {
...
+
+	status.SyncStatus.Epoch = uint64(curTs.Height())
+	timestamp := time.Unix(int64(curTs.MinTimestamp()), 0)
+	delta := time.Since(timestamp).Seconds()
+	status.SyncStatus.Behind = uint64(delta / 30)
+
+	pis, err := n.NetPeers(ctx)
+	if err != nil {
+		return status, err
+	}
```
Comment (raulk) [inline node/impl/full.go:57]: This obtains the currently _connected_ peers, which is a subset of peers with scores, right? Can we add a comment. It's not immediately obvious why we obtain peers separately, when `NetPubsubScores` already returns peers.

File: node/impl/full.go (hunk 2 above)
Comment (raulk) [reply node/impl/full.go:66]: Oh, I see. It's a bit confusing to follow due to the mixture of assignments between local vars and result struct fields.

File: node/impl/full.go (hunk 4 above)
Comment (vyzo) [reply node/impl/full.go:57]: right, NetPubScores is all peers within the retention window; will add comment.

File: node/impl/full.go (hunk 2 above)
Comment (vyzo) [reply node/impl/full.go]: done.

File: node/impl/full.go (hunk 4 above)
Comment (vyzo) [reply node/impl/full.go:58]: added comment.

File: node/impl/full.go (hunk 3 above)
Comment (vyzo) [reply node/impl/full.go:79]: crap now I have to fiddle with  DI :)

Comment (vyzo) [reply node/impl/full.go:79]: ah, stmgr seems to be able to give it to me; hooray!

Comment (vyzo) [reply node/impl/full.go]: done; there are now two separate counts -- one for blocks and once for msgs, and I use the peers listed in each topic by pubsub itself.

File: node/impl/full.go (hunk 1 above)
Comment (vyzo) [reply node/impl/full.go]: done.

**Comment by vyzo** [comment]
we briefly discussed the chain health status with @magik6k and I will add some statistics on average number of blocks in the last finality (and also last 100 epochs).

I will make this optional with a boolean argument, as it is potentially expensive and we might be calling status a lot for message sending.

**Comment by vyzo** [comment]
rebased on master

**Pull request merged**